File: CS/XpoConsoleApplication/XPComplexCustomMemberInfo.cs
using System;
using System.Collections;
using System.Collections.Generic;
using DevExpress.Xpo.Metadata;

namespace DevExpress.Xpo.Metadata {

    public class XPComplexCustomMemberInfo : XPCustomMemberInfo {
        public XPComplexCustomMemberInfo(XPClassInfo owner, string propertyName, Type propertyType, params Attribute[] attributes)
            : this(owner, propertyName, propertyType, null, false, false, attributes) { }
        public XPComplexCustomMemberInfo(XPClassInfo owner, string propertyName, XPClassInfo referenceType, params Attribute[] attributes)
            : this(owner, propertyName, null, referenceType, false, false, attributes) { }
        public XPComplexCustomMemberInfo(XPClassInfo owner, string propertyName, Type propertyType, XPClassInfo referenceType, bool nonPersistent, bool nonPublic, params Attribute[] attributes)
            : base(owner, propertyName, propertyType, referenceType, nonPersistent, nonPublic) {
            if (Equals(this.subMembersArray, XPMemberInfo.EmptyList)) {
                this.subMembersArray = new List<XPMemberInfo>();
            }
            for (int i = 0; i < attributes.Length; i++) {
                Attribute attribute = attributes[i];
                this.AddAttribute(attribute);
            }
        }
        public override bool IsStruct { get { return this.SubMembers.Count > 0; } }
        public void AddSubMember(XPComplexCustomMemberInfo memeberInfo) {
            this.SubMembers.Add(memeberInfo);
            memeberInfo.valueParent = this;
        }
        public XPComplexCustomMemberInfo AddSubMember(string propertyName, Type propertyType, params Attribute[] attributes) {
            XPComplexCustomMemberInfo memeberInfo = new XPComplexCustomMemberInfo(this.Owner, propertyName, propertyType, attributes);
            AddSubMember(memeberInfo);
            return memeberInfo;
        }
        public XPComplexCustomMemberInfo AddSubMember(string propertyName, XPClassInfo referenceType, params Attribute[] attributes) {
            XPComplexCustomMemberInfo memeberInfo = new XPComplexCustomMemberInfo(this.Owner, propertyName, referenceType, attributes);
            AddSubMember(memeberInfo);
            return memeberInfo;
        }
        public override string Name {
            get {
                if (valueParent == null) return base.Name;
                return string.Concat(this.valueParent.Name, '.', base.Name);
            }
        }
        protected override string GetDefaultMappingField() {
            if (this.valueParent == null) return base.Name;
            return this.valueParent.MappingField + base.Name;
        }
        public override object GetValue(object theObject) {
            if (IsStruct) {
                DevExpress.Xpo.Helpers.IdList idList = new DevExpress.Xpo.Helpers.IdList();
                foreach (XPMemberInfo memberInfo in this.SubMembers) {
                    if (memberInfo.IsPersistent) {
                        idList.Add(memberInfo.GetValue(theObject));
                    }
                }
                return idList;
            }
            return base.GetValue(theObject);
        }
        public override void SetValue(object theObject, object theValue) {
            if (IsStruct) {
                IList list = theValue as IList;
                if (list != null) {
                    for (int i = 0; i < list.Count; i++) {
                        XPMemberInfo memberInfo = (XPMemberInfo)this.SubMembers[i];
                        if (memberInfo.IsPersistent) {
                            memberInfo.SetValue(theObject, list[i]);
                        }
                    }
                }
                return;
            }
            base.SetValue(theObject, theValue);
        }
    }
}

File: VB/XpoConsoleApplication/XPComplexCustomMemberInfo.vb
Imports System
Imports System.Collections
Imports System.Collections.Generic

Namespace DevExpress.Xpo.Metadata

    Public Class XPComplexCustomMemberInfo
        Inherits XPCustomMemberInfo

        Public Sub New(ByVal owner As XPClassInfo, ByVal propertyName As String, ByVal propertyType As Type, ParamArray attributes As Attribute())
            Me.New(owner, propertyName, propertyType, Nothing, False, False, attributes)
        End Sub

        Public Sub New(ByVal owner As XPClassInfo, ByVal propertyName As String, ByVal referenceType As XPClassInfo, ParamArray attributes As Attribute())
            Me.New(owner, propertyName, Nothing, referenceType, False, False, attributes)
        End Sub

        Public Sub New(ByVal owner As XPClassInfo, ByVal propertyName As String, ByVal propertyType As Type, ByVal referenceType As XPClassInfo, ByVal nonPersistent As Boolean, ByVal nonPublic As Boolean, ParamArray attributes As Attribute())
            MyBase.New(owner, propertyName, propertyType, referenceType, nonPersistent, nonPublic)
            If Equals(subMembersArray, EmptyList) Then
                subMembersArray = New List(Of XPMemberInfo)()
            End If

            For i As Integer = 0 To attributes.Length - 1
                Dim attribute As Attribute = attributes(i)
                Me.AddAttribute(attribute)
            Next
        End Sub

        Public Overrides ReadOnly Property IsStruct As Boolean
            Get
                Return Me.SubMembers.Count > 0
            End Get
        End Property

        Public Sub AddSubMember(ByVal memeberInfo As XPComplexCustomMemberInfo)
            Me.SubMembers.Add(memeberInfo)
            memeberInfo.valueParent = Me
        End Sub

        Public Function AddSubMember(ByVal propertyName As String, ByVal propertyType As Type, ParamArray attributes As Attribute()) As XPComplexCustomMemberInfo
            Dim memeberInfo As XPComplexCustomMemberInfo = New XPComplexCustomMemberInfo(Owner, propertyName, propertyType, attributes)
            AddSubMember(memeberInfo)
            Return memeberInfo
        End Function

        Public Function AddSubMember(ByVal propertyName As String, ByVal referenceType As XPClassInfo, ParamArray attributes As Attribute()) As XPComplexCustomMemberInfo
            Dim memeberInfo As XPComplexCustomMemberInfo = New XPComplexCustomMemberInfo(Owner, propertyName, referenceType, attributes)
            AddSubMember(memeberInfo)
            Return memeberInfo
        End Function

        Public Overrides ReadOnly Property Name As String
            Get
                If valueParent Is Nothing Then Return MyBase.Name
                Return String.Concat(valueParent.Name, "."c, MyBase.Name)
            End Get
        End Property

        Protected Overrides Function GetDefaultMappingField() As String
            If valueParent Is Nothing Then Return MyBase.Name
            Return valueParent.MappingField & MyBase.Name
        End Function

        Public Overrides Function GetValue(ByVal theObject As Object) As Object
            If IsStruct Then
                Dim idList As Xpo.Helpers.IdList = New Xpo.Helpers.IdList()
                For Each memberInfo As XPMemberInfo In Me.SubMembers
                    If memberInfo.IsPersistent Then
                        idList.Add(memberInfo.GetValue(theObject))
                    End If
                Next

                Return idList
            End If

            Return MyBase.GetValue(theObject)
        End Function

        Public Overrides Sub SetValue(ByVal theObject As Object, ByVal theValue As Object)
            If IsStruct Then
                Dim list As IList = TryCast(theValue, IList)
                If list IsNot Nothing Then
                    For i As Integer = 0 To list.Count - 1
                        Dim memberInfo As XPMemberInfo = CType(Me.SubMembers(i), XPMemberInfo)
                        If memberInfo.IsPersistent Then
                            memberInfo.SetValue(theObject, list(i))
                        End If
                    Next
                End If

                Return
            End If

            MyBase.SetValue(theObject, theValue)
        End Sub
    End Class
End Namespace
